mllib/src/main/scala/org/apache/spark/mllib/clustering/LDA.scala

```diff
@@ -0,0 +1,472 @@
+    this.seed = seed
+    this
+  }
+
+  /**
+   * Learn an LDA model using the given dataset.
+   *
+   * @param documents  RDD of documents, where each document is represented as a vector of term
+   *                   counts plus an ID.  Document IDs must be >= 0.
+   * @return  Inferred LDA model
+   */
+  def run(documents: RDD[Document]): DistributedLDAModel = {
+    val topicSmoothing = if (this.topicSmoothing > 0) {
+      this.topicSmoothing
+    } else {
+      (50.0 / k) + 1.0
+    }
+    val termSmoothing = if (this.termSmoothing > 0) {
+      this.termSmoothing
+    } else {
+      1.1
+    }
+    var state = LDA.initialState(documents, k, topicSmoothing, termSmoothing, seed)
+    var iter = 0
+    while (iter < maxIterations) {
+      state = state.next()
+      iter += 1
+    }
+    new DistributedLDAModel(state)
+  }
+}
+
+
+object LDA {
+
+  /*
+    DEVELOPERS NOTE:
+
+    This implementation uses GraphX, where the graph is bipartite with 2 types of vertices:
+     - Document vertices
+        - indexed {0, 1, ..., numDocuments-1}
+        - Store vectors of length k (# topics).
+     - Term vertices
+        - indexed {-1, -2, ..., -vocabSize}
+        - Store vectors of length k (# topics).
+     - Edges correspond to terms appearing in documents.
+        - Edges are directed Document -> Term.
+        - Edges are partitioned by documents.
+
+    Info on EM implementation.
+     - We follow Section 2.2 from Asuncion et al., 2009.  We use some of their notation.
+     - In this implementation, there is one edge for every unique term appearing in a document,
+       i.e., for every unique (document, term) pair.
+     - Notation:
+        - N_{wkj} = count of tokens of term w currently assigned to topic k in document j
+        - N_{*} where * is missing a subscript w/k/j is the count summed over missing subscript(s)
+        - gamma_{wjk} = P(z_i = k | x_i = w, d_i = j),
+          the probability of term x_i in document d_i having topic z_i.
+     - Data graph
+        - Document vertices store N_{kj}
+        - Term vertices store N_{wk}
+        - Edges store N_{wj}.
+        - Global data N_k
+     - Algorithm
+        - Initial state:
+           - Document and term vertices store random counts N_{wk}, N_{kj}.
+        - E-step: For each (document,term) pair i, compute P(z_i | x_i, d_i).
+           - Aggregate N_k from term vertices.
+           - Compute gamma_{wjk} for each possible topic k, from each triplet.
+             using inputs N_{wk}, N_{kj}, N_k.
+        - M-step: Compute sufficient statistics for hidden parameters phi and theta
+          (counts N_{wk}, N_{kj}, N_k).
+           - Document update:
+              - N_{kj} <- sum_w N_{wj} gamma_{wjk}
+              - N_j <- sum_k N_{kj}  (only needed to output predictions)
+           - Term update:
+              - N_{wk} <- sum_j N_{wj} gamma_{wjk}
+              - N_k <- sum_w N_{wk}
+   */
+
+  /**
+   * :: DeveloperApi ::
```

Comment:
Currently, this is what I have as the contents of object LDA, to provide something for LearningStates (Gibbs, EM), to implement.

LearningState is intended to be extended by a class which represents the model/data/etc

LearningStateInitialization is intended to be extended by the companion object to create static method hooks to create a LearningState. I tried to make a single trait and let Scala see that between the companion object/class it would recognize its satisfied, but no dice (if there is a way to do this I would love to know.

```scala
object LearningAlgorithm extends Enumeration {
    type LearningAlgorithm = Value
    val Gibbs, EM = Value
  }
  type TopicCounts = Array[Int]
  case class Document(counts: Vector, id: Long)
  type DocId = Long
  type WordId = Long
  trait LearningState {
    def logPrior(): Double
    def logLikelihood(): Double
    def collectTopicTotals(): TopicCounts
    def topicDistributions(): RDD[(Long, Vector)]
    def describeTopics(maxTermsPerTopic: Long): Array[Array[(Double, String)]]
    def topicsMatrix: Matrix
    def next(): LearningState
    def vocabSize(): Long
    def k: Int
  }

  trait LearningStateInitialization {
    def initialStateFromBagOfWords(docs: RDD[Document], k: Int, alpha: Double, beta: Double, seed: Long): LearningState
    def initialStateFromEdges(edges: RDD[(WordId, DocId)], k: Int, alpha: Double, beta: Double, seed: Long): LearningState
  }
```